### DISCUSSION TITLE
Play embedded media with xelatex

### DISCUSSION THREAD

**Discussion by opensource-philosophy** [Cimbali/pympress#258] [Q&A] [answered]
First of all, thank you for making this awesome tool! I am using it regularly for my presentations at uni and am unbelievably happy with it. There is just one tiny thing I am struggling with: embedding media.

At the moment, it seems like you can't embed audio or video files with xelatex in such a way that it is played by `pympress`. The reason for this is that the `multimedia` package, for which there are examples in the README-file, does not work with xelatex, throwing error messages, and media embedded using `media9`, which _does_ work with xelatex, is not clickable in `pympress` (although it is in other `pdf` viewers such as `Okular`). Thus, I have two questions:

1. Is there a way `pympress` supports the package `media9`? It seems to be the newest, most versatile and most popular package for embedding media files in beamer.
2. If not, are you aware of any other way to embed media with xelatex in such a way that pympress recognizes it?

If I can be of any help, feel free to let me know!

**Comment by Cimbali** [comment]
If I recall correctly, the issue is the way in which the video file is embedded, which differs from across different packages. media9 wraps the movie in a flash media-player object which makes it very annoying.
- Actually executing flash is probably undesirable (lots of work, maybe shipping flash runtimes, security issues…)
- Flash itself is actually deprecated since it [reached end of life in Dec. 2020](https://www.adobe.com/products/flashplayer/end-of-life.html)
- Extracting the info from the flash object to interpret what it meant on our own is also quite a lot of work and will necessarily be incomplete or break often (any scripting for autoplay or other features will never be available).

I think the main appeal was to have a media player with the same functionality, look & feel, etc. across all platforms and PDF readers.

Alternatives are:
- `multimedia` which unfortunately does not work for you
- `movie15`, which even supports autoplay with pympress, see the example in https://github.com/Cimbali/pympress/issues/212#issuecomment-945920620 (not sure it works on xelatex?)
- lots of homemade alternatives in [this interesting question on tex.stackexchange](https://tex.stackexchange.com/questions/516029/media9-is-becoming-obsolete-dec-2020-any-alternatives-for-embedding-video-audio), which means copying some code in the header of your presentations or making your own media package from it (I find both options work well with autoplay but not quite without − that’s maybe worth digging into)

**Comment by Cimbali** [comment, marked as answer]
The `\simplemedia` approach by user Fritz on the stackexchange question works under XeLaTeX. It revealed there’s a bug in pympress auto-play I’ll get around to fixing later on. In the mean time this code works (I’ve made the original code a little bit denser):


```tex
\documentclass{beamer}

\usepackage{xkeyval}
\usepackage{zref-abspage}
\usepackage[bigfiles]{pdfbase}

% Written by Fritz -- https://tex.stackexchange.com/a/592827-- using Adobe PDF Reference, sixth edition, Nov. 2006
% Usage: \simplemedia[<options>]{<poster or text>}{<media file>}{MIME type}
\makeatletter
\define@boolkey{simplemedia}{autoplay}{\def\simplemediaautoplay{#1}}
\define@boolkey{simplemedia}{showGUI}{\def\simplemediashowGUI{#1}}
\presetkeys{simplemedia}{autoplay=false,showGUI=false}{}
\ExplSyntaxOn

\newcommand\simplemedia[4][]{%
  \setkeys{simplemedia}{#1}\leavevmode
  \pbs_pdfobj:nnn{}{fstream}{{}{#3}} % EmbeddedFiles stream
  \pbs_pdfobj:nnn{}{dict}{/Type/Filespec/F~(#3)/UF~(#3)/EF~<</F~\pbs_pdflastobj:>>} % File specification dictionary
  \pbs_pdfobj:nnn{}{dict}{/CT~(#4)/D~\pbs_pdflastobj:/N~(Media~clip~from~#3)/P~<</TF (TEMPACCESS)>>/S~/MCD} % Content type dictionary
  % XXX: /A below is incorrectly interpreted by Pympress as autoplay instead of play-on-activate and using the Additional Actions (AA) dictionary for autoplay
  \pbs_pdfobj:nnn{}{dict}{/C~\pbs_pdflastobj:/N~(Rendition~from~#3)/P~<</BE~<</C~\simplemediashowGUI~/A~\simplemediaautoplay>>~>>/S~/MR} % media clip dictionary
  \tl_set:Nx\mediarendition{\pbs_pdflastobj:}% Save ID of Rendition annotation
  \pbs_pdfobj:nnn{}{dict}{}\tl_set:Nx\renditionactionplay{\pbs_pdflastobj:}% Reserve ID for Play Rendition action
  \pbs_pdfobj:nnn{}{dict}{}\tl_set:Nx\renditionactionstop{\pbs_pdflastobj:}% Reserve ID for Stop Rendition action
  \hbox_set:Nn\l_tmpa_box{#2}% Compute size of poster
  \tl_set:Nx\l_box_wd_tl{\dim_use:N\box_wd:N\l_tmpa_box}
  \tl_set:Nx\l_box_ht_tl{\dim_use:N\box_ht:N\l_tmpa_box}
  \tl_set:Nx\l_box_dp_tl{\dim_use:N\box_dp:N\l_tmpa_box}
  \pbs_pdfxform:nnnnn{1}{1}{}{}{\l_tmpa_box}
  \tl_set:Nx\l_simplem_current_page_tl{\zref@extract{simplem@\int_use:N\g_simplem_id_int}{abspage}}
  \pbs_pdfannot:nnnn{\l_box_wd_tl}{\l_box_ht_tl}{\l_box_dp_tl}{
      /T~(Annotation~from~#3)~/Subtype~/Screen~/BS~<</W~1/S/S>>/C~[0.039216~0.039216~0.039216]/CA~1/F~4~%
      /Contents~(Embedded~media~file~#3)/NM~(simplemediascreenannot:#3)/AP~<</N~\pbs_pdflastxform:>>%
      /A~\renditionactionplay~/AA~<<\ifKV@simplemedia@autoplay~/PO~\renditionactionplay~\fi~/PC~\renditionactionstop~>>~%
      \int_compare:nT{\l_simplem_current_page_tl>0}{/P~\exp_args:Ne\pdf_pageobject_ref:n{\l_simplem_current_page_tl}}%
  }
  \pbs_pdfobj:nnn{\renditionactionplay}{dict}{/R~\mediarendition/S~/Rendition/OP~4/AN~\pbs_pdflastann:} % Rendition Action: "Play Rendition"
  \pbs_pdfobj:nnn{\renditionactionstop}{dict}{/R~\mediarendition/S~/Rendition/OP~1/AN~\pbs_pdflastann:} % Rendition Action: "Stop Rendition"
  \phantom{#2} % Display the poster
  \zref@labelbyprops{simplem@\int_use:N\g_simplem_id_int}{abspage}\int_gincr:N\g_simplem_id_int
}%
\int_new:N\g_simplem_id_int

\ExplSyntaxOff
\makeatother


\begin{document}

\begin{frame}
  \frametitle{play on click}
  \begin{center}
      \simplemedia{\includegraphics[width=5cm]{example.png}}{example.mp4}{video/mp4}
  \end{center}
\end{frame}

\begin{frame}
  \frametitle{autoplay}
  \begin{center}
      \simplemedia[autoplay=true]{\includegraphics[width=5cm]{example.png}}{example.mp4}{video/mp4}
  \end{center}
\end{frame}

\begin{frame}
  \frametitle{show GUI}
  \begin{center}
      \simplemedia[showGUI=true]{\includegraphics[width=5cm]{example.png}}{example.mp4}{video/mp4}
  \end{center}
\end{frame}

\end{document}
```

When pympress is fixed, and to have correct behaviour on other PDF players, `/A~\simplemediaautoplay` should be `/A~true`.

**Reply by sdbuch** [reply]
It may be due to an error I made during installation, but in order to get this solution to work with pympress `1.7.2_1` installed via Homebrew on MacOS, I needed to set the `GST_PLUGIN_PATH` environment variable to the GStreamer directory (e.g. for me `/opt/homebrew/lib/gstreamer-1.0`). Just mentioning in case it helps anyone else.

**Reply by sdbuch** [reply]
In case it is helpful to others, I added some very minor hacks to @Cimbali's code above, to make the functionality similar to `movie15`'s `\includemovie` -- in particular specifying a width and height for the media (rather than a box that is used to calculate the dimensions), and adding another key for whether or not to loop the media indefinitely. Here is the relevant part of the snippet:

```
% Written by Fritz -- https://tex.stackexchange.com/a/592827-- using Adobe PDF Reference, sixth edition, Nov. 2006
% Usage: \simplemedia[<options>]{<poster or text>}{<media file>}{MIME type}
% Sam: this works in pympress 1.7.2 with lualatex, info taken from
% https://github.com/Cimbali/pympress/discussions/258
\makeatletter
\define@boolkey{simplemedia}{autoplay}{\def\simplemediaautoplay{#1}}
\define@boolkey{simplemedia}{showGUI}{\def\simplemediashowGUI{#1}}
\define@choicekey*{simplemedia}{loop}[\val\nr]{true,false}{%
  \ifcase\nr\relax
    \def\simplemediarepeat{0}
  \or
    \def\simplemediarepeat{1}
  \fi
}
\presetkeys{simplemedia}{autoplay=false,showGUI=false,loop=true}{}
\ExplSyntaxOn

\newcommand\simplemedia[4][]{%
  \def\placeholderbox{%
    \vbox to #3 {\vfil%
      \hbox to #2{\hfil#4\hfil}%
      \vfil%
    }%
  }
  \setkeys{simplemedia}{#1}\leavevmode
  \pbs_pdfobj:nnn{}{fstream}{{}{#4}} % EmbeddedFiles stream
  \pbs_pdfobj:nnn{}{dict}{/Type/Filespec/F~(#4)/UF~(#4)/EF~<</F~\pbs_pdflastobj:>>} % File specification dictionary
  \pbs_pdfobj:nnn{}{dict}{/D~\pbs_pdflastobj:/N~(Media~clip~from~#4)/P~<</TF (TEMPACCESS)>>/S~/MCD} % Content type dictionary
  % XXX: /A below is incorrectly interpreted by Pympress as autoplay instead of play-on-activate and using the Additional Actions (AA) dictionary for autoplay
  \pbs_pdfobj:nnn{}{dict}{/C~\pbs_pdflastobj:/N~(Rendition~from~#4)/P~<</BE~<</C~\simplemediashowGUI~/A~\simplemediaautoplay~/RC~\simplemediarepeat>>~>>/S~/MR} % media clip dictionary
  \tl_set:Nx\mediarendition{\pbs_pdflastobj:}% Save ID of Rendition annotation
  \pbs_pdfobj:nnn{}{dict}{}\tl_set:Nx\renditionactionplay{\pbs_pdflastobj:}% Reserve ID for Play Rendition action
  \pbs_pdfobj:nnn{}{dict}{}\tl_set:Nx\renditionactionstop{\pbs_pdflastobj:}% Reserve ID for Stop Rendition action
  \hbox_set:Nn\l_tmpa_box{\placeholderbox}% Compute size of poster
  \tl_set:Nx\l_box_wd_tl{\dim_use:N\box_wd:N\l_tmpa_box}
  \tl_set:Nx\l_box_ht_tl{\dim_use:N\box_ht:N\l_tmpa_box}
  \tl_set:Nx\l_box_dp_tl{\dim_use:N\box_dp:N\l_tmpa_box}
  \pbs_pdfxform:nnnnn{1}{1}{}{}{\l_tmpa_box}
  \tl_set:Nx\l_simplem_current_page_tl{\zref@extract{simplem@\int_use:N\g_simplem_id_int}{abspage}}
  \pbs_pdfannot:nnnn{\l_box_wd_tl}{\l_box_ht_tl}{\l_box_dp_tl}{
      /T~(Annotation~from~#4)~/Subtype~/Screen~/BS~<</W~1/S/S>>/C~[0.039216~0.039216~0.039216]/CA~1/F~4~%
      /Contents~(Embedded~media~file~#4)/NM~(simplemediascreenannot:#4)/AP~<</N~\pbs_pdflastxform:>>%
      /A~\renditionactionplay~/AA~<<\ifKV@simplemedia@autoplay~/PO~\renditionactionplay~\fi~/PC~\renditionactionstop~>>~%
      \int_compare:nT{\l_simplem_current_page_tl>0}{/P~\exp_args:Ne\pdf_pageobject_ref:n{\l_simplem_current_page_tl}}%
  }
  \pbs_pdfobj:nnn{\renditionactionplay}{dict}{/R~\mediarendition/S~/Rendition/OP~4/AN~\pbs_pdflastann:} % Rendition Action: "Play Rendition"
  \pbs_pdfobj:nnn{\renditionactionstop}{dict}{/R~\mediarendition/S~/Rendition/OP~1/AN~\pbs_pdflastann:} % Rendition Action: "Stop Rendition"
  \phantom{\placeholderbox} % Display the poster
  \zref@labelbyprops{simplem@\int_use:N\g_simplem_id_int}{abspage}\int_gincr:N\g_simplem_id_int
}%
\int_new:N\g_simplem_id_int

\ExplSyntaxOff
\makeatother
```

Call as, e.g.
```
\simplemedia[autoplay=true,loop=false]{0.8\linewidth}{0.48\linewidth}{the-video.mp4}
```

**Comment by opensource-philosophy** [comment]
This is great, thank you so much for your efforts! Once the bug is fixed, could you release a new version, please?

**Comment by Cimbali** [comment]
Once it’s fixed I will update this code and put it somewhere in the documentation. However, the fix is likely going to either not be easy, or to break autoplay, so I may hold off for a while.
